kotlinx-coroutines-core/common/src/CoroutineStart.kt

```diff
@@ -1,74 +1,343 @@
      */
     @ExperimentalCoroutinesApi // Since 1.0.0, no ETA on stability
     ATOMIC,
 
     /**
-     * Immediately executes the coroutine until its first suspension point _in the current thread_ similarly to
-     * the coroutine being started using [Dispatchers.Unconfined]. However, when the coroutine is resumed from suspension
-     * it is dispatched according to the [CoroutineDispatcher] in its context.
+     * Immediately executes the coroutine until its first suspension point _in the current thread_.
+     *
+     * Starting a coroutine using [UNDISPATCHED] is similar to using [Dispatchers.Unconfined] with [DEFAULT], except:
+     * - Resumptions from later suspensions will properly use the actual dispatcher from the coroutine's context.
+     *   Only the code until the first suspension point will be executed immediately.
+     * - Even if the coroutine was cancelled already, its code will still start to be executed, similar to [ATOMIC].
+     *
+     * This set of behaviors makes [UNDISPATCHED] well-suited for cases where the coroutine has a distinct
+     * initialization phase whose side effects we want to rely on later.
+     *
+     * Example:
+     * ```
+     * runBlocking {
+     *     val channel = Channel<Int>(Channel.RENDEZVOUS)
+     *     var subscribers = 0
+     *     fun CoroutineScope.awaitTickNumber(desiredTickNumber: Int) {
+     *         launch(start = CoroutineStart.UNDISPATCHED) {
+     *             ++subscribers
+     *             try {
+     *                 for (tickNumber in channel) {
+     *                     if (tickNumber >= desiredTickNumber) {
+     *                         println("Tick number $desiredTickNumber reached")
+     *                         break
+     *                     }
+     *                 }
+     *             } finally {
+     *                 --subscribers
+     *             }
+     *         }
+     *     }
+     *     for (subscriberIndex in 1..10) {
+     *         awaitTickNumber(10 + subscriberIndex * 3)
+     *     }
+     *     // Send the current tick number every 10 milliseconds
+     *     // while there are subscribers
+     *     var i = 0
+     *     // Because of UNDISPATCHED,
+     *     // we know that the subscribers are already initialized,
+     *     // so this number is non-zero initially.
+     *     while (subscribers > 0) {
+     *         channel.trySend(++i)
+     *         delay(10.milliseconds)
+     *     }
+     * }
+     * ```
+     *
+     * Here, we implement a publisher-subscriber interaction, where [UNDISPATCHED] ensures that the
+     * subscribers do get registered before the publisher first checks if it can stop emitting values due to
+     * the lack of subscribers.
+     *
+     * **Pitfall**: unlike [Dispatchers.Unconfined] and [MainCoroutineDispatcher.immediate], nested undispatched
+     * coroutines do not form an event loop that otherwise prevents potential stack overflow in case of unlimited
```

Comment:
Do we have a resource we could link here that introduces the concept of this event loop?

Reply:
Fixed.